_drafts/schematic_review_checklist_for_firmware_engineer.md

```diff
@@ -0,0 +1,132 @@
+
+## Story 2 - Reset By Power Drain
+On the same LTE enabled e-bike, we occasionally noticed the device would become unresponsive. While we had a watchdog, for some reason the watchdog was not triggering in these cases. Once we were able to connect firmware tools (namely, a CAN dongle), we could reset the device, it would recover and work perfectly. However, we had missed designing in from the beginning a mechanism to reset the device.
+
+Unfortunately, devices stuck in this unresponsive mode were also in a low power state. Luckily, our low power mode was not that good at the time, but we ended up having to instruct our users to let bikes set for 4 days before they could recover. That's a pretty poor user experience.
+
+While we solved this in future releases by adding an NFC based reset, this is an issue that can only be fixed going forward, and not in reverse. Adding a physical way to reset the device for a human user from the beginning is a very powerful solution. How many things are fixed by "turning it off and turning it back on again"? 
+
+## Story 3 - I2C Resets
+While developing a fitness tracker, I kept running into a bug where I occasionally needed to fully power cycle a device to get the temperature sensor to start working again. The temperature sensor was connected over I2C, and digging into the issue I found that, when this was happening, the I2C SDA line was stuck low. We were able to reproduce that failure by having the MCU reset in the middle of an I2C read. Because the I2C controller on the MCU was controlling the clock, and stopped clocking, the peripheral was still driving the SDA line, waiting for the next clock signal.
+
+We had no reset line headed to the I2C device, so we could not reset the I2C device. However, we could manually clock out the remaining SCL cycles on the SCL line, and so modified our boot sequence to manually toggle the SCL line to "clock out" the remaining data, freeing up the I2C bus.
+
+However, I'd still have preferred a true reset line.
+
+## Story 4: GPIO Maps
+This is less of a war story, and more of a shoutout to good electrical engineers. Before walking into the schematic review for one project I was working on, the EE on the project published a spreadsheet of all the GPIOs for the MCU. This included their main function after reset, intended function, and any additional notes. It made it very easy as a firmware engineer to scan the list of GPIOs and ensure they were all configured properly. I'd recommend this is something that all firmware engineers ask for when looking at a schematic.
```

Comment:
This sounds great, and a forcing function for both the fw engineer and the EE to make sure you agree on the GPIO requirements. Schematics are very dense so breaking out this information makes complete sense!